### PR TITLE
Set _cronSignature to current time instead of uuid

### LINKED ISSUE #8163: cron does not run when _cronSignature is permanently not set to NOT_RUNNING
When cron starts, it adjusts the user's document to set `_cronSignature`  to the user's UID. When cron ends, it sets the value of `_cronSignature` to `NOT_RUNNING`. This is correct behaviour and is used to prevent cron running more than once at the same time.

However if something stops cron from completing (e.g., a timeout or other error), then `_cronSignature` is never reset to `NOT_RUNNING`. This prevents the user from being able to use Habitica in any way - they are not able to log in, or load the website if they are already logged in, or access their account in the mobile app.

In the past few days, we've had emails from three users affected by this. I've done a search in the database and found another 57 users affected by it, with last loggedin dates ranging from 2016-07-15 to today (for today, I checked that they weren't actually in the middle of a real cron). 16 of those users had last loggedin dates of 2016-10-19T17:31 to 2016-10-19T17:33, so they were probably affected by the CPU spike we had then ([ref for admins](https://habitica.slack.com/archives/server-health/p1476898552000013)). Another 3 users had dates of 2016-10-23T00:10 and 2016-10-23T00:11, correlating with today's server alarms ([admin ref](https://habitica.slack.com/archives/server-health/p1477181591000058)).

I've fixed all those users by setting their `_cronSignature` to `NOT_RUNNING` so they can all log in again, although I expect that some have given up trying. :(  I also made them all be resting in the inn so they won't take damage from missed Dailies.

_EDIT 2017-02-01:_
**The desired fix is that instead of setting `_cronSignature` to the User ID, set it to the current datetime, and then allow cron to run if the value is more than an hour old.**

<strike>We should adjust cron so that either it's impossible for `_cronSignature` to ever be left in an inappropriate state (might not be possible to make the code that robust) or it can recognise when it should run even though `_cronSignature` isn't set to NOT_RUNNING. Perhaps instead of setting `_cronSignature` to the User ID, we could set it to the current datetime, and then allow cron to run if the value is more than, say, an hour old. Other suggestions are welcome of course. :)</strike>

### PR DESCRIPTION
**Pull request by KyleVerma**
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Habitica_Git#Pull_Request for more info)

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
Fixes #8163

### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)

_cronSignature now set to now.getTime() instead of uuid()

created const CRON_TIMEOUT, which is set to 1hour (in ms) can be changed easily later

where the code used to make sure _cronSignature was 'NOT_RUNNING' is now an $or that also checks to see if _cronSignature is cronTimeout ms smaller than the current time

All _cronSignature time calculations are in ms since epoch time to hopefully avoid issues with timezones and end of day

[//]: # (Put User ID in here - found in Settings -> API)

----
UUID: 0036df29-70e5-4700-8e38-d7dd964a6401

### CODE REVIEW

File: website/server/middlewares/cron.js [hunk 1]
```diff
@@ -6,9 +6,10 @@ import Bluebird from 'bluebird';
 import { model as Group } from '../models/group';
 import { model as User } from '../models/user';
 import { recoverCron, cron } from '../libs/cron';
-import { v4 as uuid } from 'uuid';
 
 const daysSince = common.daysSince;
+// Wait this length of time in ms before attempting another cron
+const CRON_TIMEOUT = new Date(60 * 60 * 1000).getTime();
```
Comment (khipkin) [inline website/server/middlewares/cron.js]: Suggestion: rename variable to something more descriptive.

Ideas: CRON_TIMEOUT_WINDOW, CRON_TIMEOUT_MS, CRON_TIMEOUT_WAIT, CRON_TIMEOUT_MAX_WAIT, etc.

File: website/server/middlewares/cron.js [hunk 2]
```diff
@@ -101,12 +102,18 @@ async function cronAsync (req, res) {
       return null;
     }
 
-    let _cronSignature = uuid();
+    // set _cronSignature to current time in ms since epoch time so we can make sure to wait at least an hour before trying again even if there is a timezoned
```
Comment (khipkin) [inline website/server/middlewares/cron.js]: "even if there is a timezoned"?

Also, I suggest making this comment agnostic of the specific wait time (1 hr currently).

File: website/server/middlewares/cron.js [hunk 3]
```diff
@@ -101,12 +102,18 @@ async function cronAsync (req, res) {
...
+    let _cronSignature = now.getTime();
+    // Calculate how long ago cron must have been attempted to try again
+    let timeoutThreshold = _cronSignature - CRON_TIMEOUT;
 
     // To avoid double cron we first set _cronSignature and then check that it's not changed while processing
     let userUpdateResult = await User.update({
       _id: user._id,
-      _cronSignature: 'NOT_RUNNING', // Check that in the meantime another cron has not started
+      $or: [ // Make sure cron was successful or failed at least CRON_TIMEOUT ms ago
```
Comment (khipkin) [inline website/server/middlewares/cron.js]: Comment says CRON_TIMEOUT, but timeoutThreshold is actually used.

File: website/server/middlewares/cron.js [hunk 4]
```diff
@@ -101,12 +102,18 @@ async function cronAsync (req, res) {
       return null;
     }
 
-    let _cronSignature = uuid();
+    // set _cronSignature to current time in ms since epoch time so we can make sure to wait at least an hour before trying again even if there is a timezoned
+    let _cronSignature = now.getTime();
+    // Calculate how long ago cron must have been attempted to try again
+    let timeoutThreshold = _cronSignature - CRON_TIMEOUT;
```
Comment (khipkin) [inline website/server/middlewares/cron.js]: Suggestion: rename variable to something more descriptive. Can't think of any suggestions right now.

File: website/server/middlewares/cron.js (hunk 1 above)
Comment (KyleVerma) [reply website/server/middlewares/cron.js]: Gonna go with CRON_TIMEOUT_WAIT if nobody has a better suggestion, it seems most accurate in my mind to what the constant is actually storing

File: website/server/middlewares/cron.js (hunk 4 above)
Comment (KyleVerma) [reply website/server/middlewares/cron.js]: how does cronRetryTime sound to you as a more accurate/descriptive variable name?

Comment (khipkin) [reply website/server/middlewares/cron.js]: @KyleVerma Ohh that sounds perfect!

**Comment by KyleVerma** [comment]
Am I supposed to change the stub for the failing test or is changing tests handled by admins?

I can't figure out how to change stubs in sinon.js in a way that will work nicely with the new update call so if I'm supposed to change it can someone point me in the direction of something that will be more helpful than sinon.js documentation? Thanks.

**Comment by paglias** [comment]
@KyleVerma The test is failing because the stub uses `.withArgs()` https://github.com/HabitRPG/habitica/blob/develop/test/api/v3/unit/middlewares/cronMiddleware.js#L238 which means it'll return the value in `.returns()` only when the arguments passed to `User.update` match the ones in `.withArgs`. To fix it you should make sure `withArgs` is passed the new arguments that are now passed to User.update() which are now

```
       _id: user._id,
       _cronSignature: 'NOT_RUNNING', // Check that in the meantime another cron has not started
       $or: [ // Make sure last cron was successful or failed before cronRetryTime
         {_cronSignature: 'NOT_RUNNING'},
         {_cronSignature: {$lt: cronRetryTime}},
       ],
```

**Comment by KyleVerma** [comment]
Sorry if I'm being dense but how do I handle cronRetryTime being a variable? Is there somewhere else where the stub is being called that I can't find where I can put in 'fake' times or something?

**Comment by paglias** [comment]
@KyleVerma no worries. You can use sinon.js matchers http://sinonjs.org/releases/v2.0.0/matchers/
that make it possible to match a value without knowing it completely.

Something like this could work

```
withArgs({
       _id: user._id,
       $or: [ // Make sure last cron was successful or failed before cronRetryTime
         {_cronSignature: 'NOT_RUNNING'},
         {_cronSignature: {$lt: sinon.match.number}},
       ],
});
```

**Comment by KyleVerma** [comment]
Worked like a charm thanks!

Should a test be created for the new functionality?

I was thinking something along the lines of:
- Simulate a failed cron; either just set _cronSignature to some time or force cron to timeout somehow
- Try an update call less than CRON_TIMEOUT_WAIT after the failure
- Try an update call more than CRON_TIMEOUT_WAIT after the failure (unless we want tests that take an hour I'm guessing with fake timestamps)

**Comment by paglias** [comment]
I think it's enough to have these two tests:

- Set cron_signature (using User.update()) to more than 1 hour ago and verify that calling `cron()` re-execute it
- Set cron to 59 minutes ago and verify that calling cron() doesn't re-execute it

**Comment by KyleVerma** [comment]
A few questions concerning the commit I just pushed:

- What is the naming convention for new unit tests?
- What are the standards/guidelines for creating/throwing new errors in unit tests? (concerning the first test I added)
- Is there a nicer way to handle setting cronSignature to 1 hour ago? (Ideally it would somehow reference the variable in cron.js so if it's changed there these tests don't also need to be changed)
- In the first test should I check the error message to verify that cron is failing for the right reason? Or check something else before resolving?

**Comment by Alys** [comment]
@KyleVerma I'm really sorry, we never answered your questions. :( Do you still need answers for any/all of them? Or is this ready for us to review?

**Comment by KyleVerma** [comment]
I think I have tests that correctly test the functionality added, so a code review couldn't hurt. However my questions are still potentially necessary since my tests might not be up to standards without them.

**Comment by SabreCat** [comment]
Naming convention: The naming convention for unit tests is pretty loose vs. the more regimented one we have for integration tests. You could name this one with `.test.js` instead if you'd like, but this is fine.

Creating/throwing errors: That's interesting--I don't know that we've structured any other tests this way (without an explicit `expect`). @TheHollidayInn or @paglias, any thoughts?

Constants: We would typically put magic numbers like that in a constant at the top of the test file and reference it that way (so at least it's clear where you need to update if the functionality does change). The alternative would be to export a constant somewhere and import it into both the middleware and the test, but I don't know that we use that pattern outside of `/website/common`.

Check error message: Ideally we should check that the right error is thrown, yes. And that could be an answer for the second question--go ahead and set up the test with an `expect` that the cron will error out a specific way.

**Comment by TheHollidayInn** [comment]
You should be able to `await` and expectation which will resolve promises. Here is an example: https://github.com/HabitRPG/habitica/blob/develop/test/api/v3/unit/libs/applePayments.test.js#L52

**Comment by SabreCat** [comment]
@KyleVerma, are you still working on this? If not, let us know and we can open this up for someone else to tackle.

**Comment by KyleVerma** [comment]
So sorry for the delay I had some work deadlines and an currently on vacation. Should be able to wrap this up when I get home.

**Comment by SabreCat** [comment]
@KyleVerma New ping. Back from vacation yet?

**Comment by KyleVerma** [comment]
Yes I just got back recently and was planning to do some work on this today.

**Comment by SabreCat** [comment]
Some linting errors, shouldn't be too hard to fix!

**Comment by KyleVerma** [comment]
Gonna try and get to them by the end of the day :)

Is it a problem that one of the tests I added takes 1.5s to run?

**Comment by KyleVerma** [comment]
Two tests are failing due to timeouts now. Did I accidentally cause this? I'll look into it further soon but didn't want to waste time if someone could easily look at the tests and determine that I'm not the culprit. Thanks.

**Comment by Alys** [comment]
@KyleVerma I've restarted the tests. If it times out again we'll look into it in more detail, but from a quick glance it didn't seem to be related to your code.

**Comment by KyleVerma** [comment]
Is there anything that still needs to be done on my end for this issue?

**Comment by SabreCat** [comment]
No, we mainly just set this aside because we've been prioritizing PRs related to the recent frontend redesign! We'll take a look at this soon.

**Comment by KyleVerma** [comment]
Makes sense

**Comment by SabreCat** [comment]
@KyleVerma Thank you for your effort (and patience!). I've awarded you your first contributor tier--welcome to the ranks of the Blacksmiths!

**Comment by KyleVerma** [comment]
Awesome, Thanks!

**Pull request merged**